src/Store.ts

```diff
@@ -0,0 +1,392 @@
+	new (args?: {}): any;
+	createSubclass?: (arg: any[]) => any;
+	prototype: any;
+}
+
+export interface StoreArgs {
+	Model?: NewableStoreModel;
+}
+
+export default class Store<T> extends Evented implements dstore.Collection<T>, Hash<any> {
+	[ index: string ]: any;
+	/**
+	 * Creates an object, throws an error if the object already exists
+	 *
+	 * @param object The object to store.
+	 * @param directives Additional directives for creating objects.
+	 *
+	 * @return The object that was stored, with any changes that were made by
+	 * the storage system (like generated id)
+	 */
+	add: (object: any, directives?: dstore.PutDirectives) => any;
+
+	/**
+	 * Indicates if the events should automatically be fired for put, add, remove
+	 * method calls. Stores may wish to explicitly fire events, to control when
+	 * and which event is fired.
+	 */
+	autoEmitEvents: boolean = true;
+
+	fetch: (args?: dstore.FetchArgs) => dstore.FetchPromise<T>;
+
+	Filter: typeof Filter = Filter;
+
+	/**
+	 * Retrieves an object by its identity
+	 *
+	 * @param id The identity to use to lookup the object
+	 * @returns The object in the store that matches the given id.
+	 */
+	get: (id: dstore.StoreItem) => Promise<T> | void;
+
+	/**
+	 * Indicates the property to use as the identity property. The values of this
+	 * property should be unique.
+	 */
+	idProperty: string = 'id';
+
+	/**
+	 * This should be a entity (like a class/constructor) with a 'prototype' property that will be
+	 * used as the prototype for all objects returned from this store. One can set
+	 * this to the Model from dmodel/Model to return Model objects, or leave this
+	 * to null if you don't want any methods to decorate the returned
+	 * objects (this can improve performance by avoiding prototype setting),
+	 */
+	Model: NewableStoreModel;
+
+	/**
+	 * One can provide a parsing function that will permit the parsing of the data. By
+	 * default we assume the provide data is a simple JavaScript array that requires
+	 * no parsing (subclass stores may provide their own default parse function)
+	 */
+	parse: (...args: any[]) => any;
+
+	/**
+	 * Stores an object
+	 *
+	 * @param object The object to store.
+	 * @param directives Additional directives for storing objects.
+	 * @returns The object that was stored, with any changes that were made by
+	 * the storage system (like generated id)
+	 */
+	put: (object: any, directives?: dstore.PutDirectives) => Promise<any>;
+
+	/**
+	 * Indicates if client-side query engine filtering should (if the store property is true)
+	 * access object properties through the get() function (enabling querying by
+	 * computed properties), or if it should (by setting this to false) use direct/raw
+	 * property access (which may more closely follow database querying style).
+	 */
+	queryAccessors: boolean =  true;
+
+	/**
+	 * Deletes an object by its identity
+	 *
+	 * @param id The identity to use to delete the object
+	 */
+	remove: (id: any) => Promise<Object>;
```

Comment:
`remove` currently can return either an object or a boolean in dstore 1.x (see Memory and Rest).  We might want to tighten that up in 2.0.  Returning a boolean seems particularly pointless when `remove` already returns a promise (if it fails, it'll reject), so maybe `Promise<T | void>`?.